homeassistant/components/romy/vacuum.py

```diff
@@ -0,0 +1,288 @@
+import json
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components.vacuum import (
+    PLATFORM_SCHEMA,
+    VacuumEntity,
+    VacuumEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_NAME, CONF_PASSWORD, CONF_PORT
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .utils import async_query
+
+_LOGGER = logging.getLogger(__name__)
+
+ICON = "mdi:robot-vacuum"
+PLATFORM = "romy"
+
+FAN_SPEED_NONE = "Default"
+FAN_SPEED_NORMAL = "Normal"
+FAN_SPEED_SILENT = "Silent"
+FAN_SPEED_INTENSIVE = "Intensive"
+FAN_SPEED_SUPER_SILENT = "Super_Silent"
+FAN_SPEED_HIGH = "High"
+FAN_SPEED_AUTO = "Auto"
+
+FAN_SPEEDS: list[str] = [
+    FAN_SPEED_NONE,
+    FAN_SPEED_NORMAL,
+    FAN_SPEED_SILENT,
+    FAN_SPEED_INTENSIVE,
+    FAN_SPEED_SUPER_SILENT,
+    FAN_SPEED_HIGH,
+    FAN_SPEED_AUTO,
+]
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PORT, default=8080): cv.port,
+        vol.Optional(CONF_PASSWORD): vol.All(str, vol.Length(8)),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+# Commonly supported features
+SUPPORT_ROMY_ROBOT = (
+    VacuumEntityFeature.BATTERY
+    | VacuumEntityFeature.PAUSE
+    | VacuumEntityFeature.RETURN_HOME
+    | VacuumEntityFeature.SEND_COMMAND
+    | VacuumEntityFeature.STATUS
+    | VacuumEntityFeature.STOP
+    | VacuumEntityFeature.TURN_OFF
+    | VacuumEntityFeature.TURN_ON
+    | VacuumEntityFeature.FAN_SPEED
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up ROMY vacuum with config entry."""
+    host = config_entry.data[CONF_HOST]
+    port = config_entry.data[CONF_PORT]
+    name = config_entry.data[CONF_NAME]
+    password = config_entry.data.get(CONF_PASSWORD, "")
+    unique_id = ""
+    model = ""
+    firmware = ""
+
+    ret, response = await async_query(hass, host, port, "get/robot_id")
+    if ret:
+        status = json.loads(response)
+        unique_id = status["unique_id"]
+        model = status["model"]
+        firmware = status["firmware"]
+    else:
+        _LOGGER.error("Error fetching unique_id resp: %s", response)
+
+    device_info = {
+        "manufacturer": "ROMY",
+        "model": model,
+        "sw_version": firmware,
+        "identifiers": {"serial": unique_id},
+    }
+
+    _LOGGER.info(
+        "Adding ROMY vacuum robot with IP: %s PORT:%s, NAME: %s, UNIQUE_ID:%s, PASS:%s",
+        host,
+        port,
+        name,
+        unique_id,
+        password,
+    )
+    romy_vacuum_entity = RomyRobot(host, port, name, unique_id, device_info, password)
+
+    entities = [romy_vacuum_entity]
+    async_add_entities(entities, True)
```

Comment:
If you fail to get the information you need from that request, you shouldn't just continue setting it up anyway, the setup should fail.